### ISSUE TITLE
apd: wrong result with UNION, EXCEPT, DISTINCT, GROUP BY 

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#6877]
### What version of Materialize are you using?

```
v0.7.4-dev (519d7ab07)
```

### What was the issue?

The ```UNION EXCEPT``` and the rest of the operations does not work properly with the APD data type.

### Is the issue reproducible? If so, please provide reproduction instructions.

```
CREATE TABLE insertions (f1 APD) ;
CREATE TABLE deletions (f1 APD) ;
INSERT INTO insertions VALUES ('1'::APD);
INSERT INTO deletions VALUES ('1'::APD);
SELECT * FROM insertions EXCEPT ALL SELECT * FROM deletions;

 f1 
----
 1 <- wrong, no rows should have been returned.
(1 row)
```

```
materialize=> 
materialize=> DROP TABLE IF EXISTS t1;
DROP TABLE
materialize=> CREATE TABLE t1 (f1 APD);
CREATE TABLE
materialize=> INSERT INTO t1 VALUES ('1'::APD);
INSERT 0 1
materialize=> INSERT INTO t1 VALUES ('1'::APD);
INSERT 0 1
materialize=> SELECT DISTINCT f1 FROM t1;
 f1 
----
 1
 1 <- wrong, duplicates should have been removed
(2 rows)
```

**Comment by sploiselle** [comment]
Seems like there's also some kind of race condition (?) in the distinct code because I'm occasionally getting the right answer.

```
materialize=>
materialize=> DROP TABLE IF EXISTS t1;
DROP TABLE
materialize=> CREATE TABLE t1 (f1 APD);
CREATE TABLE
materialize=> INSERT INTO t1 VALUES ('1'::APD);
INSERT 0 1
materialize=> INSERT INTO t1 VALUES ('1'::APD);
INSERT 0 1
materialize=> SELECT DISTINCT f1 FROM t1;
 f1
----
 1
 1
(2 rows)

materialize=> SELECT DISTINCT f1 FROM t1;
 f1
----
 1
 1
(2 rows)


materialize=> DROP TABLE IF EXISTS t1;
DROP TABLE
materialize=> CREATE TABLE t1 (f1 APD);
CREATE TABLE
materialize=> INSERT INTO t1 VALUES ('1'::APD);
INSERT 0 1
materialize=> INSERT INTO t1 VALUES ('1'::APD);
INSERT 0 1
materialize=> SELECT DISTINCT f1 FROM t1;
 f1
----
 1
(1 row)
```

**Comment by sploiselle** [comment]
Have a fix for the underlying issue in MaterializeInc/rust-dec#35

**Comment by philip-stoev** [comment]
Unfortunately even after I forced rust-dec to be at ```9ec4426``` :+1: 

```
diff --git a/Cargo.toml b/Cargo.toml
index a5c06c558..a0138ba99 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -68,3 +68,6 @@ rustc-demangle = { opt-level = 3 }
 # it's unlikely we're going to get interactive access to a debugger in
 # production installations, but we still want useful crash reports.
 debug = 1
+
+[patch.crates-io]
+dec = { git = 'https://github.com/MaterializeInc/rust-dec', branch = 'master' }
```

and my Cargo.lock became:

```
[[package]]
name = "dec"
version = "0.4.0"
source = "git+https://github.com/MaterializeInc/rust-dec?branch=master#9ec4426c6345de51504f7967e34ec58c3521e858"
dependencies = [
 "decnumber-sys",
 "libc",
 "serde",
 "static_assertions",
]

[[package]]
name = "decnumber-sys"
version = "0.1.5"
source = "git+https://github.com/MaterializeInc/rust-dec?branch=master#9ec4426c6345de51504f7967e34ec58c3521e858"
dependencies = [
 "cc",
 "libc",
]
```

the issue remains repeatable. The original test case reproduces a problem reliably with ```--workers 1```:

```
materialize=>  DROP TABLE IF EXISTS insertions CASCADE ;
DROP TABLE
materialize=>   DROP TABLE IF EXISTS deletions CASCADE ;
DROP TABLE
materialize=> 
materialize=> CREATE TABLE insertions (f1 APD) ;
CREATE TABLE
materialize=> CREATE TABLE deletions (f1 APD) ;
CREATE TABLE
materialize=> INSERT INTO insertions VALUES ('1'::APD);
INSERT 0 1
materialize=> INSERT INTO deletions VALUES ('1'::APD);
INSERT 0 1
materialize=> SELECT * FROM insertions EXCEPT ALL SELECT * FROM deletions;
 f1 
----
 1
(1 row)
```
With workers = N it seems that it is reproducible every N-th time or thereabout.

Furthermore, a new test case became available that is also reproducing a wrong result with workers = N. It seems to me that it is the same thing, it is just that slightly more rows are involved:

```
materialize=>  DROP TABLE IF EXISTS insertions CASCADE ;
DROP TABLE
materialize=>   DROP TABLE IF EXISTS deletions CASCADE ;
DROP TABLE
materialize=>   CREATE TABLE insertions (f1 APD) ;
CREATE TABLE
materialize=>   CREATE TABLE deletions (f1 APD) ;
CREATE TABLE
materialize=> 
materialize=> 
materialize=> INSERT INTO deletions VALUES  ( CONCAT( '+00386.' , 3 , 8 , 4 )::APD );
INSERT 0 1
materialize=> INSERT INTO insertions VALUES ( '+0.1 '::APD );
INSERT 0 1
materialize=> INSERT INTO deletions VALUES ( '+0.3831552 '::APD ) , ( '+0.1 '::APD );
INSERT 0 2
materialize=> 
materialize=> 
materialize=> SELECT * FROM insertions EXCEPT ALL SELECT * FROM deletions;
 f1  
-----
 0.1
(1 row)
```

**Comment by sploiselle** [comment]
@philip-stoev tysm for continuing to test this; turns out it's caused by an uninitialized byte in our current deserialization from struct to datum; introducing a method that will let us fix this in MaterializeInc/rust-dec#37. tested the change out locally and the issues are resolvable.

**Comment by philip-stoev** [comment]
Yes, thank you, using your branch all the test cases in this ticket now return correct results.
